maps.py

```diff
@@ -1,49 +1,47 @@
 from abc import ABC, abstractmethod
-from pathlib import Path
-from typing import Generator, Optional
-from urllib.parse import urlparse
+from typing import Generator, Optional, Type, Tuple
 
-from pygeotile.tile import Tile
+from darkgeotile import BaseTile, get_Tile
 
-from utils import ImageFormat
-from utils import get_random_tile
+from pyproj import Proj
 
 
 class Map(ABC):
-    def __init__(self, *args, **kwargs):
-        raise NotImplementedError
+    Tile: Type[BaseTile]
+
+    # Attributes for overwriting:
+    projection: Proj
+    projection_bounds: Optional[Tuple[float, float, float, float]] = None
 
     @staticmethod
     @abstractmethod
-    def get_urls_gen(tile: Tile) -> Generator[str, None, None]:
+    def get_urls_gen(tile) -> Generator[str, None, None]:
         raise NotImplementedError
 
     @staticmethod
     def get_timeout():
         return 0
 
-    tiles_format = None  # should be overridden
+    @classmethod
+    def get_tile_gen(cls, bbox: Tuple[float, float, float, float], zoom: int) -> Generator[Type[BaseTile], None, None]:
+        tms_x_s, tms_y_s = zip(*[tile.tms for tile in cls.get_corner_tiles(bbox, zoom)])
+
+        for x in range(min(tms_x_s), max(tms_x_s) + 1):
+            for y in range(min(tms_y_s), max(tms_y_s) + 1):
+                yield cls.Tile.from_tms(x, y, zoom)
 
     @classmethod
-    def guess_tiles_format(cls) -> Optional[ImageFormat]:
-        # language=rst
-        """
-        Guessing tiles format of :py:class:`Map` subclass by random tile url
-        (method called if subclass doesn't override `cls.tiles_format` attribute)
-        """
-        for random_tile_url in cls.get_urls_gen(get_random_tile()):
-            suffix = Path(urlparse(random_tile_url)[2]).suffix
-            proper_formats = [img_f for img_f in ImageFormat if suffix in img_f.possible_suffixes]
-            if proper_formats:
-                return proper_formats[0]
+    def get_corner_tiles(cls, bbox, zoom):
+        return tuple(cls.Tile.for_xy(x, y, zoom) for x in bbox[::2] for y in bbox[1::2])
 
     def __init_subclass__(cls, **kwargs):
-        if cls.tiles_format is None:
-            expected_tiles_format = cls.guess_tiles_format()
-            if expected_tiles_format:
-                cls.tiles_format = expected_tiles_format
-            else:
-                raise Exception("can't guess tiles format")
+        if cls.projection is None:
+            raise Exception('unknown coordinate reference system')
+        elif not isinstance(cls.projection, Proj):
+            cls.projection = Proj(cls.projection)
+
+        cls.Tile = get_Tile(cls.projection, cls.projection_bounds)
```

Comment:
Try to rename "get_Tile" function. I think "get_tile_class" looks more correctly than now. And add annotation of return value for this function.